#####################################################################################################################
############################################## Set up ###############################################################
#####################################################################################################################

# Import packages and custom wrapper functions
source("src/bulk_wrappers.R")


# other libraries
library(fgsea)
library(AnnotationDbi)
library(eulerr)
library(DOSE)
library(proxy)
library(reshape2)
library(ComplexHeatmap)
library(ggdendro)
library(pheatmap)
library(gtable)


#####################################################################################################################
########################################### Prepare preranked list ##################################################
#####################################################################################################################





# Import DESEQ2 results
degDESeqH7m <- read_csv("huHDAC7_KD/results/DGE/degDESeqH7m.csv")


# GSEA requirements

# Based on:
#   
#   https://stephenturner.github.io/deseq-to-fgsea/
#   
#   http://evomics.org/wp-content/uploads/2018/10/RNASeqWorkFlow-1.html

#Remove potential duplicates entries based on gene names
#(they can be unprocessed transcript or alternative mapping to neighbor locus)
#We arbitrarily choose the entry that has the highest baseMean expression.
#It is advised to use gene symbols for input in GSEA
#I checked the difference between gene symbol and gene name and there is no difference.
#So using the gene_name column shoud be ok.
degDESeqH7.nodup <- degDESeqH7m %>% 
  group_by(gene_name_H7m) %>% 
  top_n(n = 1, wt = baseMean_H7m) %>%
  ungroup()



#Generate the ranked file by selecting the gene name and the statistics variables and remove NAs
RNK_H7 <- degDESeqH7.nodup %>%
  mutate(Gene_Name = gene_name_H7m) %>%
  dplyr::select(Gene_Name, stat_H7m) %>%
  dplyr::filter(is.na(Gene_Name)==FALSE) %>%
  arrange(-stat_H7m)

#Convert variant results to a vector:
RNK_H7Vector <- deframe(RNK_H7)



#####################################################################################################################
######################################### Prepare FOXO KO genesets ##################################################
#####################################################################################################################

########################
###### import data #####
########################

# CHIP from Foxo papers
affy_ensembl = c("affy_mouse430_2", "ensembl_gene_id", "external_gene_name")
affy_mouse430_2 <- getBM(attributes= affy_ensembl, mart= ensembl, values = "*", uniqueRows=T) %>%
  filter(affy_mouse430_2 != "")


#first FOXO dataset
#Import output DEG files generated by GEO2R and modify logFC column with name of comparison
#DEGs found on GEOR for the FOXO1.3 KO are inverted compared to DEGs in papers, so I revert by changing the sign of the fold change.
DEGsfoxo1.3 <- read.delim("huHDAC7_KD/data/foxo1.3.TregKO.txt") %>% 
  plyr::rename(c("logFC" = "foxo1.3.lfc")) %>%
  plyr::rename(c("P.Value" = "foxo1.3.pval")) %>%
  mutate(foxo1.3.lfc = -foxo1.3.lfc) %>%
  left_join(affy_mouse430_2, by = c("ID" = "affy_mouse430_2"))


#Mapping to ensembl IDs is largely incomplete, 
#try to recover the remaining probeIDs from Illumina website
lost <- DEGsfoxo1.3 %>% 
  dplyr::filter(is.na(ensembl_gene_id))

#upload list of retrieved IDs from affymetrix website
retrievedIDs <- read.delim("huHDAC7_KD/data/lost-retrieved-from-affy.tsv") %>%
  mutate(mapping = ifelse(grepl("---", Ensembl), "unmapped", ifelse(grepl("///", Ensembl), "multiple mapping", "unique mapping"))) %>%
  group_by(mapping) %>%
  summarise(count = n())

#count how many probe IDs we could retreive.
kable(retrievedIDs, caption = "Retreived probe IDs mapping to Ensemble IDs from the Affymetrix website")

#Import the second FOXO dataset
DEGsfoxo1 <- read.delim("huHDAC7_KD/data/foxo1.TregKO.txt") %>% 
  plyr::rename(c("logFC" = "foxo1.lfc")) %>%
  plyr::rename(c("P.Value" = "foxo1.pval")) 

############################################
###### merge and define genesets data ######
############################################


#Merge the 2 datasets using the gene IDs
#Remove genes with no ensembl ID
merged.foxo <- DEGsfoxo1.3 %>% 
  dplyr::select(Gene.symbol, ID, foxo1.3.lfc, foxo1.3.pval, ensembl_gene_id, external_gene_name) %>%
  inner_join(DEGsfoxo1 %>% dplyr::select(foxo1.lfc, foxo1.pval, ID), by = "ID") %>%
  dplyr::filter(!is.na(ensembl_gene_id)) %>%
  dplyr::mutate(sign = ifelse(foxo1.3.pval <0.05 | foxo1.pval <0.0001, "diff", "non diff")) %>%
  dplyr::mutate(corr = ifelse(abs(foxo1.3.lfc/foxo1.lfc) >4, "foxo3dep", ifelse(abs(foxo1.3.lfc/foxo1.lfc) <2, "foxo1dep", "int")))


ggplot() +
  geom_point(data = merged.foxo %>% dplyr::filter(sign != "diff"), aes(x= foxo1.lfc, y = foxo1.3.lfc), color = "grey50" ,alpha = 0.4) +
  geom_point(data = merged.foxo %>% dplyr::filter(sign == "diff"), aes(x= foxo1.lfc, y = foxo1.3.lfc, color = corr) ,alpha = 0.4) +
  ggtitle("Correlation between Foxo1 KO and Foxo1&3 double KO datasets") +
  geom_abline(slope = 3, line) +
  geom_abline(slope = -3) +
  scale_color_viridis(end = 0.8, discrete = T) +
  theme_mrl(0.5)



merged.foxo.geneset.info <- merged.foxo %>%
  mutate(Genesets = ifelse(corr == "foxo1dep" & sign == "diff" 
                           & foxo1.3.lfc*foxo1.lfc >0 & foxo1.lfc >0 & foxo1.3.pval < 0.05, "FOXO1 UP",
                           ifelse(corr == "foxo1dep" & sign == "diff" 
                                  & foxo1.3.lfc*foxo1.lfc >0 & foxo1.lfc <0 & foxo1.3.pval < 0.05, "FOXO1 DN",
                                  ifelse(corr == "foxo3dep" & sign == "diff" 
                                         & foxo1.3.lfc >0 & foxo1.3.pval < 0.05 & foxo1.lfc < 0.25, "FOXO3 UP",
                                         ifelse(corr == "foxo3dep" & sign == "diff" 
                                                & foxo1.3.lfc <0 & foxo1.3.pval < 0.05 & foxo1.lfc > -0.25, "FOXO3 DN", NA)))))



ggplot() +
  geom_point(data = merged.foxo.geneset.info, aes(x= foxo1.lfc, y = foxo1.3.lfc), color = "grey50", alpha = 0.4) +
  geom_point(data = merged.foxo.geneset.info, aes(x= foxo1.lfc, y = foxo1.3.lfc, color = Genesets), alpha = 0.4) +
  scale_color_viridis(end = 0.8, discrete = T) +
  theme_mrl(0.5) +
  labs(y = "FOXO1 FOXO3 double KO log2FC", x = "FOXO1 KO log2FC") +
  theme(aspect.ratio = 1, 
        axis.line = element_blank(),
        panel.border = element_rect(colour = "black", fill=NA, size=0.5)) +
  lims(x = c(-5,5), y = c(-5,5))

ggsave("huHDAC7_KD/results/figs/Foxos_genesets.pdf", width = 5, height = 3)

# Generate genesets for foxo1 and foxo3 dependent genes.

foxo1upGeneSet <- merged.foxo %>%
  dplyr::filter(corr == "foxo1dep" 
                & sign == "diff" 
                & foxo1.3.lfc*foxo1.lfc >0 
                & foxo1.lfc >0 
                & foxo1.3.pval < 0.05) %>%
  dplyr::pull(ensembl_gene_id) %>%
  convertMouseGeneList() %>%
  dplyr::filter(Gene.type == "protein_coding") %>%
  dplyr::pull(Gene.name) %>%
  toupper()

foxo1dnGeneSet <- merged.foxo %>%
  dplyr::filter(corr == "foxo1dep" 
                & sign == "diff" 
                & foxo1.3.lfc*foxo1.lfc >0 
                & foxo1.lfc <0 
                & foxo1.3.pval < 0.05) %>%
  dplyr::pull(ensembl_gene_id) %>%
  convertMouseGeneList() %>%
  dplyr::filter(Gene.type == "protein_coding") %>%
  dplyr::pull(Gene.name) %>%
  toupper()

foxo3upGeneSet <- merged.foxo %>%
  dplyr::filter(corr == "foxo3dep" & sign == "diff" 
                & foxo1.3.lfc >0 & foxo1.3.pval < 0.05 & foxo1.lfc < 0.25) %>%
  dplyr::pull(ensembl_gene_id) %>%
  convertMouseGeneList() %>%
  dplyr::filter(Gene.type == "protein_coding") %>%
  dplyr::pull(Gene.name) %>%
  toupper()

foxo3dnGeneSet <- merged.foxo %>%
  dplyr::filter(corr == "foxo3dep" & sign == "diff" 
                & foxo1.3.lfc <0 & foxo1.3.pval < 0.05 & foxo1.lfc > -0.25) %>%
  dplyr::pull(ensembl_gene_id) %>%
  convertMouseGeneList() %>%
  dplyr::filter(Gene.type == "protein_coding") %>%
  dplyr::pull(Gene.name) %>%
  toupper()

#####################################################################################################################
######################################### Prepare FOXO CA genesets ##################################################
#####################################################################################################################



# Use activated Treg RNAseq dataset from 2016 Nature paper on FOXO1
# Can be downloaded at https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4978705/bin/NIHMS805312-supplement-Data_Table.xlsx
# and reformatted to a tab delimited file, removing the top part to get the column headers in the first row
aTregDESeq <- read.delim("huHDAC7_KD/data/aTregNaturePaper.txt", stringsAsFactors = F) %>%
  separate(ensemblID, into = c("EnsemblID"), by = ".")

#Constitutively active foxo genesets
CA_rescue_Up <- aTregDESeq %>%
  dplyr::filter(regulation.CA_rescue. == "Upregulated") %>%
  dplyr::pull(EnsemblID) %>%
  convertMouseGeneList() %>%
  dplyr::filter(Gene.type == "protein_coding") %>%
  dplyr::pull(Gene.name) %>%
  toupper()

CA_rescue_Dn <- aTregDESeq %>%
  dplyr::filter(regulation.CA_rescue. == "Downregulated") %>%
  dplyr::pull(EnsemblID) %>%
  convertMouseGeneList() %>%
  dplyr::filter(Gene.type == "protein_coding") %>%
  dplyr::pull(Gene.name) %>%
  toupper()




#####################################################################################################################
########################################### GSEA on FOXO sets #######################################################
#####################################################################################################################


############################
## Assemble FOXO genesets ##
############################


foxoGeneSets <- list(foxo1upGeneSet, 
                     foxo1dnGeneSet, 
                     foxo3upGeneSet, 
                     foxo3dnGeneSet,
                     CA_rescue_Up,
                     CA_rescue_Dn)

names(foxoGeneSets) <- c("Foxo1 KO UP", 
                         "Foxo1 KO DN", 
                         "Foxo3 KO UP", 
                         "Foxo3 KO DN",
                         "Foxo1 CA UP",
                         "Foxo1 CA DN")


#########################
## Analysis with fgsea ##
#########################

fgseaVarFoxos <- fgsea(pathways=foxoGeneSets, stats=RNK_H7Vector, nperm=50000)

fgseaVarFoxos <- fgseaVarFoxos %>% mutate(Overlap = (str_count(leadingEdge,",")+1)/size*100)


enrichment.plots <- lapply(c("Foxo1 KO UP", "Foxo1 KO DN", 
                             "Foxo3 KO UP", "Foxo3 KO DN",
                             "Foxo1 CA UP", "Foxo1 CA DN"),
                           function(x) plotEnrichment(foxoGeneSets[[x]],RNK_H7Vector))

########################
## Analysis with DOSE ##
########################

# Same analysis with the DOSE wrapper that provides a nice interface for making enrichment plots
# the gseaplot2 from enrichplot package not available on R3.5 so I sourced the function.

detach("package:dplyr", unload=TRUE)
library(clusterProfiler)
library(dplyr)
library(enrichplot)
source("src/gseaplot2_PPAmod.R")




test <- list_to_df(foxoGeneSets) %>% 
  unnest() %>% 
  plyr::rename(c("name" = "Term")) %>% 
  plyr::rename(c("list.element" = "Gene"))



gse <- GSEA(geneList = RNK_H7Vector, nPerm = 50000, minGSSize = 40, maxGSSize = 5000, TERM2GENE = test, pvalueCutoff = 1)

gseaplot2(gse, geneSetID = 1, title = gse$Description[1], color = "black") 

ggsave("huHDAC7_KD/results/figs/gseaplots.pdf", width = 6, height = 5.5)




gseResult <- gse@result %>%
  separate(leading_edge, into = c("tags","list","signal"), sep = ",") %>%
  mutate(Overlap = as.integer(gsub("[^0-9.-]", "", tags))) %>%
  mutate(list = as.integer(gsub("[^0-9.-]", "", list))) %>%
  mutate(signal = as.integer(gsub("[^0-9.-]", "", signal)))


Foxo1UPcore <- gseResult %>% 
  dplyr::filter(ID == "Foxo1 KO UP") %>% 
  dplyr::select(core_enrichment) %>% 
  unnest(core_enrichment = strsplit(core_enrichment, "/")) %>%
  pull(core_enrichment)



ggplot(gseResult, aes(x =ID, y = NES, size = Overlap, fill = -log10(pvalue))) +
  geom_point(shape = 21, stroke = 0.1, color = "black") +
  geom_hline(yintercept = 0, linetype = "dashed") +
  scale_y_continuous(limits = c(-max(abs(gseResult$NES)),max(abs(gseResult$NES)))) +
  coord_flip() +
  scale_size_continuous(range = c(2,7)) +
  scale_fill_viridis(begin = 0.2,limits = c(1.30104,NA)) +
  labs(x="", y = "Normalized enrichment score") +
  theme_mrl(0.6) +
  theme(axis.title.y=element_blank(),
        axis.ticks = element_line(),
        axis.line = element_blank(),
        axis.text = element_text(colour = "black"),
        legend.position = "right",
        panel.border = element_rect(colour = "black", fill=NA, size=0.5),
        aspect.ratio = 1,
        legend.key.size = unit(0.3, "cm"))

ggsave("huHDAC7_KD/results/figs/gseaplots2.pdf", width = 5, height = 3)
